### ISSUE TITLE
trying to add sanitizers

### ISSUE THREAD

**Issue by svenevs** [wjakob/nanogui#271]
(feel free to close, I put some time into this and wanted to make it available, but it's far from PR ready.  I am very impressed by the sanitizers, and will probably revisit this at some point regardless)

I learned of the `clang` sanitizers recently, and they helped me find a race condition in my code.  The race condition was actually from something very similar to how nanogui does things in `common.cpp`

```cpp
static bool mainloop_active = false;
void mainloop(int refresh) { 
    // create a `refresh_thread` that uses `mainloop_active`
    // ...
    // do the main loop on the main thread with `mainloop_active`
}
```

Though relatively minor, this is technically a race condition.  The solution is to just `#include <atomic>` and use `static std::atomic<bool> mainloop_active{false};`, but I couldn't actually reproduce this error.  I thought maybe I didn't understand, but then I added some `std::cout` in both the main loop and `refresh_thread`, which is _definitely_ a race condition (and also reported in my own project when I did this).  But something about the NanoGUI build system seems to disable the sanitizers.

My understanding is you're a `clang` fan, and thought maybe you'd be interested in these.  The diff I'm working with right now (needs more testing on Linux) is as follows, the changes to `common.cpp` _should_ be reporting a race condition but are not.  Is there anything obvious about why the sanitizers don't seem to actually do anything?

If you apply the diff here, you'd build with `cmake .. -DNANOGUI_DEV_SANITIZE_THREAD=ON`.

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21ab22e..8082906 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -544,4 +544,78 @@ else()
 
 endif()
 
+# Developer sanitizer checks can be incorporated; example
+#     cmake .. -DNANOGUI_DEV_SANITIZE=ON
+# to sanitize address among many others (if not apple clang)
+#     cmake .. -DNANOGUI_DEV_SANITIZE_THREAD=ON
+# to sanitize threads.  Threads should be clean, but this
+# can be a useful sanity check.
+#
+# See https://clang.llvm.org/docs/UsersManual.html#controlling-code-generation
+if (NANOGUI_DEV_SANITIZE OR NANOGUI_DEV_SANITIZE_THREAD)
+  if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    # All of the examples suggest -O1 -g
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g")
+    if(CMAKE_CXX_FLAGS MATCHES "-O[0-4]")
+      string(REGEX REPLACE "-O[0-4]" "-O1" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+    else()
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O1")
+    endif()
+
+    # Sanitize thread and address cannot be used at the same time, default to
+    # sanitizing address, but allow thread bypass
+    if (NANOGUI_DEV_SANITIZE_THREAD)
+      # https://clang.llvm.org/docs/ThreadSanitizer.html
+      list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "thread")
+    else()
+      # https://clang.llvm.org/docs/AddressSanitizer.html
+      list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "address")
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-omit-frame-pointer")
+    endif()
+
+    # Add the control flow integrity check if using flto
+    # NOTE: kept getting segfault with this
+    # if (LTO_LINKER_FLAGS MATCHES "flto")
+    #   # https://clang.llvm.org/docs/ControlFlowIntegrity.html
+    #   list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "cfi")
+    # endif()
+    # Apple Clang does not support dataflow, or memory
+    # and cannot link against undefined or safe-stack
+    execute_process(
+      COMMAND ${CMAKE_CXX_COMPILER} -v
+      # clang -v outputs on stderr
+      ERROR_VARIABLE NANOGUI_CLANG_VERSION
+    )
+    if (NOT NANOGUI_CLANG_VERSION MATCHES "Apple")
+      # https://clang.llvm.org/docs/MemorySanitizer.html
+      list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "memory")
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-omit-frame-pointer -fsanitize-memory-track-origins=2")
+
+      # https://clang.llvm.org/docs/DataFlowSanitizer.html
+      # seems to require manual instrumentation, not used
+      # list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "dataflow")
+
+      # https://clang.llvm.org/docs/SafeStack.html
+      list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "safe-stack")
+
+      # https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
+      # NOTE: this probably needs more attention
+      list(APPEND NANOGUI_DEV_SANITIZE_CHECKS "undefined")
+    endif()
+    # add the sanitizers (compile and link)
+    foreach (sanitizer ${NANOGUI_DEV_SANITIZE_CHECKS})
+      message(STATUS "NANOGUI_DEV: adding sanitizer ${sanitizer}")
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=${sanitizer}")
+      set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fsanitize=${sanitizer}")
+      if (NANOGUI_BUILD_SHARED)
+        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -fsanitize=${sanitizer}")
+      else()
+        set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} -fsanitize=${sanitizer}")
+      endif()
+    endforeach()
+  else()
+    message(FATAL_ERROR "NANOGUI_DEV_SANITIZER_CHECKS can only be used with Clang")
+  endif()
+endif()
+
 # vim: set et ts=2 sw=2 ft=cmake nospell:
diff --git a/src/common.cpp b/src/common.cpp
index 80b6d59..928b999 100644
--- a/src/common.cpp
+++ b/src/common.cpp
@@ -60,7 +60,7 @@ void init() {
 }
 
 static bool mainloop_active = false;
-
+#include <iostream>
 void mainloop(int refresh) {
     if (mainloop_active)
         throw std::runtime_error("Main loop is already running!");
@@ -78,6 +78,7 @@ void mainloop(int refresh) {
                 std::chrono::milliseconds time(refresh);
                 while (mainloop_active) {
                     std::this_thread::sleep_for(time);
+                    std::cout << "I am a race condition" << std::endl;
                     glfwPostEmptyEvent();
                 }
             }
@@ -87,6 +88,7 @@ void mainloop(int refresh) {
     try {
         while (mainloop_active) {
             int numScreens = 0;
+            std::cout << "Mainlooooop" << std::endl;
             for (auto kv : __nanogui_screens) {
                 Screen *screen = kv.second;
                 if (!screen->visible()) {

```

**Comment by wjakob** [comment]
The sanititizers are great-- I use the memory/address sanitizers a lot. In this case, I don't see a race condition as writes to booleans are always atomic (and the main loop is only supposed to be executed by the main thread).

**Comment by svenevs** [comment]
I had modeled something very similar to the `nanogui::mainloop` with virtually an identical setup (in fact, my `mainloop` was internally calling `nanogui::mainloop` after some setup to spawn a data harvesting thread), and the sanitizers got mad.  In practice it shouldn't matter, but in theory it's to prevent the compiler from doing bad optimizations (?).

In all seriousness, I don't think NanoGUI needs to change this.  It was simply my reference point (coupled with `std::cout`).  I'll probably revisit this in January when I have some more free time to actually do some more reading on them / engineer some explicit cases that the sanitizers should be catching.  I'd also probably change the dev interface to just be `-DNANOGUI_DEV_SANITIZE=thread` and just run one sanitizer at a time.  Some of them are complementary though, I'm not sure.  I closed it because the above isn't worthy of a PR, but it's definitely not an issue.
